V4 Programmatic Open edX Certificates

caesar2164 opened the pull request:
- Add renderers/elements.py that draws element on page given standardized inputs. Available elements are:
  - Rectangle
  - Line
  - Image
  - Text
  - Formatted Text
- Draw certificate template programmatically
- Draw Instructor Signature Blocks programmatically apart from signature image which must be provided as an image

@stvstnfrd - here's the real PR, with squashed commits!

Commit:
V4 Programmatic Open edX Certificates

- Add renderers/elements.py that draws element on page given standardized inputs. Available elements are:
  - Rectangle
  - Line
  - Image
  - Text (can contain context variables)
- Render certificate template programmatically
- Render Instructor Signature Blocks programmatically apart from signature image which must be provided as an image
- Render course certificate content programmatically from cert-data

Commit:
Add settings defaults outside env.json conditional

Also:
- Ensure GEN_TMP_DIR exists before trying to use it
- Add default for instructor signatures

stvstnfrd commented on cert-data.yml:
```diff
@@ -12,6 +187,19 @@ edX/DemoX_v3/Demo_Course_v3:
+  instructors:
+    - instructor:
+        <<: *instructor_1
+        name: "<b>M. Guillaume Hébert</b>"
```
Is this a reference of which I'm unaware?

caesar2164 replied:
It's a play on the "Head Concierge of the Grand Budapest Hotel" :) 

Just wanted some silly name that would also have UTF8 chars in it, and for which I could safely make a bogus signature...

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        width: 600
+        height: 14
+    - text:
+        string: "<b>{student_name}</b>"
```
We need to document all possible string formats.

Can you add these to the README?

We should probably have an entire section devoted to V4.

caesar2164 replied:
yeppers, good call.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+  instructor_template:
+    # ordered list of steps to render the instructor signature block, bottom to top
+    - text:
+        string: "school"
```
Especially since we actually do have interpolated strings throughout this file, I would prefer we use single quotes for all non-interpolated strings.

caesar2164 replied:
ok, I'll fix this throughout...

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+  TEMPLATEFILE: False
+  instructors:
+    - instructor: &instructor_1
+        x_position: 475
```
We should document each of these with something intuitive like "top right" or "far right, 3rd from top".

caesar2164 replied:
Where would you like me to put this documentation? In this file?

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        font_size: 10
+        leading: 10
+        text_color: [.18,.18,.18]
+        alignment: TA_RIGHT
```
Documentation should acknowledge the relationship between this and reportlab.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        string: "school"
+        font_size: 10
+        leading: 10
+        text_color: [.18,.18,.18]
```
We should document this.

Tangentially, how easy would it be to write this with with hex in the YAML and then convert to this (decimal?) notation on the backend?

caesar2164 replied:
It looks like I could:

`import matplotlib.colors as colors`

and then do:

`colors.hex2color('#2e2e2e')        #==> (.18,.18,.18)`

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        width: 250
+        height: 12
+    - image:
+        file: "signature_file"
```
Shouldn't this be an actual file path?

caesar2164 replied:
no this is the template, so it takes the file path provided in the course description further down in cert-data

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        file: "signature_file"
+        width: 200
+        height: 50
+        x_offset: 50
```
What is this again?

caesar2164 replied:
space from the left to make up the difference in width with text displayed.

stvstnfrd commented:
We need to document the font licenses in the README.

stvstnfrd commented on settings.py:
```diff
@@ -90,9 +103,9 @@
-    QUEUE_NAME = ENV_TOKENS.get('QUEUE_NAME', 'test-pull')
-    QUEUE_URL = ENV_TOKENS.get('QUEUE_URL', 'https://stage-xqueue.edx.org')
+    TMP_GEN_DIR = ENV_TOKENS.get('TMP_GEN_DIR', TMP_GEN_DIR)
+    QUEUE_NAME = ENV_TOKENS.get('QUEUE_NAME', QUEUE_NAME)
```
Wait, does this mean previously `QUEUE_NAME` hadn't been defined?

It looks like if `env.json` existed _but_ it did not contain a value for `QUEUE_NAME`, an exception would be thrown.

caesar2164 replied:
???

if env.json doesn't have anything it should take the default it gets passed, no?

stvstnfrd commented on settings.py:
```diff
@@ -90,9 +103,9 @@
-    QUEUE_URL = ENV_TOKENS.get('QUEUE_URL', 'https://stage-xqueue.edx.org')
+    TMP_GEN_DIR = ENV_TOKENS.get('TMP_GEN_DIR', TMP_GEN_DIR)
+    QUEUE_NAME = ENV_TOKENS.get('QUEUE_NAME', QUEUE_NAME)
+    QUEUE_URL = ENV_TOKENS.get('QUEUE_URL', QUEUE_URL)
```
Same thing here.

stvstnfrd commented on settings.py:
```diff
@@ -49,6 +56,12 @@
+TMP_GEN_DIR = '/tmp/certificates/'
+
+QUEUE_NAME = 'test-pull'
+QUEUE_URL = 'https://stage-xqueue.edx.org'
```
I know this was already here, but we should definitely not be defaulting to a potentially real/existing URL.

Let's change this to `https://xqueue.example.com`.

Commit:
Add default course and v4 demo course

- default course has template/layout settings
- V4 demo course shows how to use the default course tweaks to render multiple courses

stvstnfrd commented:
Per F2F, I think we want to license the `cert-data.yml` separately under a more developer-permissive license like BSD, MIT, etc. This will allow adopters to extend this file in their private template directory, without being required to license as AGPL derived.

We should note this in both the `README` and within the header of the `cert-data.yml`.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+  instructor_template:
+    # ordered list of steps to render the instructor signature block, bottom to top
+    - text:
+        string: 'school'
```
Change all instances of `string` within `instructor_template` to `key`.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,153 @@
+        for element, attributes in ordered_element_copy.iteritems():
+            x_offset = attributes.get('x_offset', 0)
+            if element == "text":
+                attributes['string'] = instructor_attributes.get(attributes.get("string", ""), "")
```
Change `attributes.get("string", "")` to `attributes.get("key", "")`.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+    # ordered list of steps to render the instructor signature block, bottom to top
+    - text:
+        string: 'school'
+        font_size: 10
```
Note in docs (maybe just in the header of this file?) that all positions and sizes are listed in points.

stvstnfrd commented:
TODO: Can we translate from decimal colors to hex?

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        string: 'school'
+        font_size: 10
+        leading: 10
+        text_color: [.18,.18,.18]
```
http://www.reportlab.com/apis/reportlab/2.4/lib.html#reportlab.lib.colors.HexColor

stvstnfrd commented on cert-data.yml:
```diff
@@ -12,6 +187,19 @@ edX/DemoX_v3/Demo_Course_v3:
+    - instructor:
+        <<: *instructor_1
+        name: '<b>M. Guillaume Hébert</b>'
+        title: 'Head Concierge & Lifestyle Manager'
```
Add second title line to this example too.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        width: 250
+        height: 12
+    - text:
+        string: 'title_2'
```
Maybe `supplemental_title` or some variation thereof.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        width: 250
+        height: 12
+    - image:
+        file: 'signature_file'
```
Also change to `key`.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,153 @@
+            elif element == "image":
+                attributes['x'] = x_position + x_offset
+                attributes['y'] = y_position - 10
+                file_name = attributes.get("file", "")
```
`file` to `key`

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        y: 54
+        width: 680
+        height: 488
+        stroke: 1
```
Remove this and instead check `if stroke_width`.

``` python
if stroke_width or stroke_color:
    stroke_width = stroke_width or 1
    stroke_color = stroke_color or '#000000'
    stroke = 1
```

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        width: 680
+        height: 488
+        stroke: 1
+        fill: 1
```
Remove this and instead check `if fill_color`.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        x_position: 475
+        y_position: 140
+    - instructor: &instructor_2
+        x_position: 475
```
Change to `x`.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        y_position: 140
+    - instructor: &instructor_2
+        x_position: 475
+        y_position: 220
```
Change to `y`.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,153 @@
+
+
+def draw_instructor_element(certificateGen, instructor_attributes, instructor_template, canvas):
+    x_position = instructor_attributes.get("x_position", 0)
```
Change to `.get('x', 0)`.

caesar2164 commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        height: 12
+    - image:
+        file: 'signature_file'
+        width: 200
```
set this to 250 like the other ones

caesar2164 commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+    - image:
+        file: 'signature_file'
+        width: 200
+        height: 50
```
remove and use `img.getSize()` from report lab

caesar2164 commented on cert-data.yml:
```diff
@@ -1,4 +1,179 @@
+        file: 'signature_file'
+        width: 200
+        height: 50
+        x_offset: 50
```
remove

caesar2164 commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,153 @@
+                attributes['x'] = x_position + x_offset
+                attributes['y'] = y_position - 10
+                file_name = attributes.get("file", "")
+                attributes['file'] = instructor_attributes.get(file_name, "")
```
also check `instructor_attributes.get('y_offset', 0)`

caesar2164 commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,153 @@
+            x_offset = attributes.get('x_offset', 0)
+            if element == "text":
+                attributes['string'] = instructor_attributes.get(attributes.get("string", ""), "")
+                attributes['x'] = x_position + x_offset
```
remove

caesar2164 commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,153 @@
+    for ordered_element in instructor_template:
+        ordered_element_copy = copy.deepcopy(ordered_element)
+        for element, attributes in ordered_element_copy.iteritems():
+            x_offset = attributes.get('x_offset', 0)
```
remove

caesar2164 commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,153 @@
+                else:
+                    continue
+            elif element == "image":
+                attributes['x'] = x_position + x_offset
```
remove

caesar2164 commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,153 @@
+                    continue
+            elif element == "image":
+                attributes['x'] = x_position + x_offset
+                attributes['y'] = y_position - 10
```
move as first line in `if attributes['file']:` and calculate image height using: http://stackoverflow.com/questions/5327670/image-aspect-ratio-using-reportlab-in-python

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,194 @@
 ---
+### 4_programmatic Default course
+Default/OpenEdX/Certificate: &cert_defaults
```
`certificate_defaults`

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,194 @@
+  VERSION: 4_programmatic
+  TEMPLATEFILE: False
+  certificate_theme:
+    # ordered list of steps to render the certificate theme, back to front
```
"typically won't change course to course"
- borders, logos, etc.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,194 @@
 ---
+### 4_programmatic Default course
```
add note that all sizes are in POINTS

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,194 @@
+              alignment: right
+              width: 250
+              height: 12
+          - text: &sup_title
```
supplemental

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,194 @@
+        template:
+          # Ordered list of steps to render an instructor signature block, bottom to top
+          # (each element is located relative to the previous one)
+          - text: &school
```
Prefix with `instructor_default_`?

stvstnfrd replied:
Add this prefix to each of the default instructor blocks.

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,194 @@
+          - text: &school
+              key: 'school_string'
+              font_size: 10
+              leading: 10
```
Maybe add a comment to the first appearance of this word?

stvstnfrd commented on cert-data.yml:
```diff
@@ -1,4 +1,194 @@
+              key: 'signature_file'
+              width: 250
+    # Convenient instructor signature block locations
+    - instructor: &instructor_1
```
Add a description of where this is; bottom-right, 2nd from left..

stvstnfrd replied:
maybe a picture or ascii art instead

stvstnfrd commented on cert-data.yml:
```diff
@@ -12,6 +202,60 @@ edX/DemoX_v3/Demo_Course_v3:
+        <<: *instructor_1
+        name_string: '<b>M. Guillaume Hébert</b>'
+        title_string: 'Head Concierge & Lifestyle Manager'
+        sup_title_string: ''
```
maybe add a non-empty value

stvstnfrd commented on gen_cert.py:
```diff
@@ -12,6 +12,7 @@
 
 from reportlab.platypus import Paragraph
 from PyPDF2 import PdfFileWriter, PdfFileReader
+from reportlab.lib import utils
```
Alphabetize this import.

stvstnfrd commented on gen_cert.py:
```diff
@@ -41,6 +42,7 @@
 from openedx_certificates.renderers.util import autoscale_text
 from openedx_certificates.renderers.util import font_for_string
 from openedx_certificates.renderers.util import WIDTH_LANDSCAPE_PAGE_IN_POINTS
+from openedx_certificates.renderers.elements import draw_template_element
```
Alphabetize this import.

stvstnfrd commented on gen_cert.py:
```diff
@@ -166,18 +170,25 @@ def interstitial_factory():
         # If TEMPLATEFILE is set in cert-data.yml, this value has top priority.
         # Else if a value is passed in to the constructor (eg, from xqueue), it is used,
         # Else, the filename is calculated from the version and course_id.
+
```
Nit: Not of fan of additional newlines :)

stvstnfrd commented on gen_cert.py:
```diff
@@ -166,18 +170,25 @@ def interstitial_factory():
-            log.critical("I/O error ({0}): {1} opening {2}".format(e.errno, e.strerror, template_pdf_filename))
-            raise
+
+        if template_pdf is not False:
```
Document why this is like this.

stvstnfrd commented on gen_cert.py:
```diff
@@ -166,18 +170,25 @@ def interstitial_factory():
-            raise
+
+        if template_pdf is not False:
+            template_prefix = '{0}/v{1}-cert-templates'.format(TEMPLATE_DIR, self.template_version)
```
Double quotes for interpolated strings.

stvstnfrd replied:
While we're editing this string template, let's use named kwargs instead of positional options.

stvstnfrd commented on gen_cert.py:
```diff
@@ -166,18 +170,25 @@ def interstitial_factory():
+
+        if template_pdf is not False:
+            template_prefix = '{0}/v{1}-cert-templates'.format(TEMPLATE_DIR, self.template_version)
+            template_pdf_filename = "{0}/certificate-template-{1}-{2}.pdf".format(
```
While we're editing this string template, let's use named kwargs instead of positional options.

stvstnfrd commented on gen_cert.py:
```diff
@@ -166,18 +170,25 @@ def interstitial_factory():
+                self.course,
+            )
+            if template_pdf:
+                template_pdf_filename = "{0}/{1}".format(template_prefix, template_pdf)
```
While we're editing this string template, let's use named kwargs instead of positional options.

stvstnfrd commented on gen_cert.py:
```diff
@@ -166,18 +170,25 @@ def interstitial_factory():
+                if 'verified' in template_pdf:
+                    self.template_type = 'verified'
+            try:
+                self.template_pdf = PdfFileReader(file(template_pdf_filename, "rb"))
```
single quotes

stvstnfrd commented on gen_cert.py:
```diff
@@ -166,18 +170,25 @@ def interstitial_factory():
+            try:
+                self.template_pdf = PdfFileReader(file(template_pdf_filename, "rb"))
+            except IOError as e:
+                log.critical("I/O error ({0}): {1} opening {2}".format(e.errno, e.strerror, template_pdf_filename))
```
For log statements, use lazy interpolation:

``` python
log.critical("I/O error (%s): %s opening %s", e.errno, e.strerror, template_pdf_filename)
```

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+        RETURNS (download_uuid, verify_uuid, download_url)
+        """
+
+        verify_me_p = self.cert_data.get('VERIFY', True)
```
verify_paragraph?

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+
+        # This file is overlaid on the template certificate
+        overlay_pdf_buffer = StringIO.StringIO()
+        PAGE = canvas.Canvas(overlay_pdf_buffer, pagesize=landscape(A4))
```
no caps? not actually a constant.

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+        overlay_pdf_buffer = StringIO.StringIO()
+        PAGE = canvas.Canvas(overlay_pdf_buffer, pagesize=landscape(A4))
+
+        WIDTH, HEIGHT = landscape(A4)  # Width and Height of landscape canvas (in points)
```
width, height

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+        for instructor in instructors:
+            for element, attributes in instructor.iteritems():
+                x_position = attributes.get('x', 0)
+                y_position = attributes.get('y', 0)
```
`attributes['y']`: This should break if not set.

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+        instructors = self.cert_data.get('instructors', [])
+        for instructor in instructors:
+            for element, attributes in instructor.iteritems():
+                x_position = attributes.get('x', 0)
```
`attributes['y']`: This should break if not set.

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+                for template_item in attributes['template']:
+                    for item_type, item_properties in template_item.iteritems():
+                        if item_type == 'text':
+                            key = item_properties['key']
```
``` python
value = item_properties.get('string')
if not value:
    key = item_properties['key']
    value = attributes[key]
item_properties['string'] = value
```

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+                            key = item_properties['key']
+                            item_properties['string'] = attributes.get(key, '')
+                            if not item_properties['string']:
+                                continue
```
Above code block replaces 2021-2024

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+                                continue
+                            item_properties['x'] = x_position
+                            item_properties['y'] = y_position
+                            y_position += item_properties.get('height', 0)
```
Break if unset.

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+                            key = item_properties['key']
+                            item_properties['file'] = attributes.get(key, '')
+                            if not item_properties['file']:
+                                continue
```
Replace 2034->2037 with:

``` python
value = item_properties.get('file')
if not value:
    key = item_properties['key']
    value = attributes[key]
item_properties['file'] = value
```

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+                                continue
+
+                            filepath = os.path.join(TEMPLATE_DIR, item_properties['file'])
+                            img = utils.ImageReader(filepath)
```
image

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+
+                            filepath = os.path.join(TEMPLATE_DIR, item_properties['file'])
+                            img = utils.ImageReader(filepath)
+                            img_width, img_height = img.getSize()
```
image_width, image_height

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+                            img = utils.ImageReader(filepath)
+                            img_width, img_height = img.getSize()
+
+                            aspect = img_height / float(img_width)
```
`aspect_ratio`

stvstnfrd commented on gen_cert.py:
```diff
@@ -1853,3 +1865,230 @@ def _generate_v3_dynamic_certificate(
+
+                            aspect = img_height / float(img_width)
+                            img_width = item_properties.get('width', img_width)
+                            img_height = img_width * aspect
```
`image_height = int(image_width * aspect_ratio)`

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+# -*- coding: utf-8 -*-
+
+import logging
```
alphabetize all import stanzas.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+    return alignment
+
+
+def draw_rect(_certificateGen, attributes, canvas, _context):
```
`draw_rectangle` please :)

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+
+def draw_rect(_certificateGen, attributes, canvas, _context):
+    stroke = 0
+    stroke_color = attributes.get('stroke_color', 0)
```
`stroke_color = attributes.get('stroke_color')`

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+def draw_rect(_certificateGen, attributes, canvas, _context):
+    stroke = 0
+    stroke_color = attributes.get('stroke_color', 0)
+    stroke_width = attributes.get('stroke_width', 0)
```
Also remove default `0` here.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+        canvas.setStrokeColor(HexColor(stroke_color))
+
+    fill = 0
+    fill_color = attributes.get('fill_color', 0)
```
Remove default here.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+            attributes.get('y', 0),
+            attributes.get('width', 0),
+            attributes.get('height', 0),
+            stroke=attributes.get('stroke', 1),
```
`stroke=stroke`,

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+            attributes.get('width', 0),
+            attributes.get('height', 0),
+            stroke=attributes.get('stroke', 1),
+            fill=attributes.get('fill', 1),
```
`fill=fill,`

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+
+    if stroke or fill:
+        canvas.rect(
+            attributes.get('x', 0),
```
for x, y, height, width, grab unsafely from dict
`attributes['y']`, etc.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+
+def draw_line(_certificateGen, attributes, canvas, _context):
+    stroke = 0
+    stroke_color = attributes.get('stroke_color', 0)
```
Drop defaults here and on next line.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+    if stroke_color or stroke_width:
+        stroke_color = stroke_color or '#000000'
+        stroke_width = stroke_width or 1
+        stroke = 1
```
Delete this.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+        canvas.setLineWidth(stroke_width)
+        canvas.setStrokeColor(HexColor(stroke_color))
+
+    if stroke:
```
Delete this line, move block up into previous `if`

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+
+    if stroke:
+        canvas.line(
+            attributes.get('x_start', 0),
```
Grab unsafely.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+
+
+def draw_image(_certificateGen, attributes, canvas, _context):
+    image_file = os.path.join(TEMPLATE_DIR, attributes.get('file', '/'))
```
attributes['file']

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+    image_file = os.path.join(TEMPLATE_DIR, attributes.get('file', '/'))
+    canvas.drawImage(
+        image_file,
+        attributes.get('x', 0),
```
unsafe plz

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+
+
+def draw_text(certificateGen, attributes, canvas, context):
+    string = unicode(attributes.get('string', ''))
```
Make unsafe.

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+
+    fontSize = attributes.get('font_size', 12)
+    leading = attributes.get('leading', 12)
+    textColor = attributes.get('text_color', 0)
```
default to '#000000'

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+    fontSize = attributes.get('font_size', 12)
+    leading = attributes.get('leading', 12)
+    textColor = attributes.get('text_color', 0)
+    height = attributes.get('height', 0)
```
Make unsafe.
- height
- width
- x_position
- y_position

stvstnfrd commented on openedx_certificates/renderers/elements.py:
```diff
@@ -0,0 +1,143 @@
+    paragraph.drawOn(canvas, x_position, y_position)
+
+
+element_options = {
```
ELEMENT_OPTIONS

stvstnfrd commented on settings.py:
```diff
@@ -34,12 +34,18 @@
+# Directory in which to store Logs
+LOG_DIR = '/var/tmp'
+# Local Logging Level
+local_loglevel = 'INFO'
```
probably capitalized?

stvstnfrd commented:
Be sure to list font licenses in README.

stvstnfrd commented:
+1

stvstnfrd committed:
Remove trailing whitespace

stvstnfrd committed:
Clarify values vs. sizes

stvstnfrd committed:
Rename variables within nested loops for clarify
